storage/watcher.go

```diff
@@ -69,6 +75,39 @@ type WatchResponse struct {
 
 	// CompactRevision is set when the watcher is cancelled due to compaction.
 	CompactRevision int64
+
+	// State reports the status of watch.
+	State WatchState
+}
+
+// WatchState reports the status of watch.
+type WatchState struct {
+	// StartRevision is the revision that watch started on.
+	StartRevision int64
+
+	// Progress is the progress status of watcher.
+	Progress WatchProgress
+}
+
+type WatchProgress int
+
+const (
+	// Watched is default value when watch events happen.
+	Watched WatchProgress = iota
+	// WatchOngoing represents the watcher that is still waiting for events.
+	WatchOngoing
+	// WatchDeleted means the watcher got deleted after compaction.
+	WatchDeleted
```

Comment:
i do not think we need watchDeleted. user will receive the compacted event out of the chan anyway.